### DISCUSSION TITLE
RTK Query: results from different query arguments returned on error

### DISCUSSION THREAD

**Discussion by dumaron** [reduxjs/redux-toolkit#3349] [Q&A] [answered]
Hello everyone 🖖 
I have a strange issue that I believe should not occur by reading the documentation. I'm building a slice with RTK Query with `fakeBaseQuery` since the real - and only - endpoint is implemented by performing a graphql request.

The issue appears when the endpoint returns an error: while the status correctly is being set to "rejected" the data from the previous query, which was performed with different query arguments, is shown.

You can see the issue reproduced on [this codesandbox](https://codesandbox.io/s/optimistic-https-6dc60z?file=/src/index.tsx:329-342): if you click on "set error" button you'll see that while the query has changed argument, the data from the previous one is still shown.

I played with several combinations of `provideTags`, `invalidateTags`, and other things, but nothing seems to work for me. Is it a bug, or am I missing something?

**Comment by EskiMojo14** [comment, marked as answer]
see [Frequently Used Query Hook Return Values](https://redux-toolkit.js.org/rtk-query/usage/queries#frequently-used-query-hook-return-values) for the difference between `data` and `currentData`:

- `data` - The latest returned result regardless of hook arg, if present.
- `currentData` - The latest returned result for the current hook arg, if present.

you can see that your [sandbox](https://codesandbox.io/s/sweet-violet-dgd85k) behaves the way you expected when using `currentData`.

**Reply by EskiMojo14** [reply]
also, your queryFn should never return **both** data and error, it should only ever return an object containing one of the keys. { data, meta? } or { error, meta? } but never { data, error, meta? }

**Reply by dumaron** [reply]
Many thanks @EskiMojo14 , as I expected the answer was simple and yet invisible to my eyes.
